### PR TITLE
Additional OAuth2 providers

### PR DESCRIPTION
**Pull request by willemvd**
This is a followup on #679 (this PR is based on the merged #679 master) which only included GitHub as an OAuth2 provider.

For now the following providers are implemented (incl signin image and tip text in admin/auths/new):
Google+
GitLab 
Bitbucket
Twitter
Facebook
Dropbox
OpenID Connect

For GitLab and Github it is now possible to change the default urls to custom urls, this means you can connect them to your own private/local/on-premise/enterprise installation

I will add and test new providers on request in this PR (or on gitter?), list of possible providers can be found at : https://github.com/markbates/goth/#supported-providers

### CODE REVIEW

**Comment by lunny** [comment]
build failed.

**Comment by willemvd** [comment]
fixed

File: models/login_source.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"code.gitea.io/gitea/modules/auth/pam"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/auth/oauth2"
+	"sort"
```
Comment (appleboy) [inline models/login_source.go]: Move local package to top

File: modules/auth/oauth2/oauth2.go [hunk 2]
```diff
@@ -15,6 +15,14 @@ import (
...
 	"path/filepath"
 	"github.com/markbates/goth/providers/github"
+	"github.com/markbates/goth/providers/gplus"
+	"github.com/markbates/goth/providers/gitlab"
+	"github.com/markbates/goth/providers/bitbucket"
+	"github.com/markbates/goth/providers/twitter"
+	"github.com/markbates/goth/providers/facebook"
+	"github.com/markbates/goth/providers/dropbox"
+	"github.com/markbates/goth/providers/openidConnect"
+	"math"
```
Comment (appleboy) [inline modules/auth/oauth2/oauth2.go]: See https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#styleguide

Comment (willemvd) [reply modules/auth/oauth2/oauth2.go]: done, changed to goimports style

File: models/login_source.go (hunk 1 above)
Comment (willemvd) [reply models/login_source.go]: done, changed to goimports style

**Comment by appleboy** [comment]
@willemvd build fails

**Comment by willemvd** [comment]
Build failure over a file I didn't change and cannot be found at that location? 😕 
modules/helper.go -> golint failure but file does not exist

**Comment by appleboy** [comment]
@willemvd Please rebase the latest version of master. See https://github.com/go-gitea/gitea/pull/1017

**Comment by willemvd** [comment]
Not sure what happend , but rebase seems to have messed up my PR 
See duplicate commits in my history 😕 
Not sure what will happen when this gets merged... maybe create a new PR ?

**Comment by willemvd** [comment]
@lunny @appleboy any idea ? (see my previous comment) & when I do a merge locally on master with this PR , history will be messed up with duplicate entries

**Comment by lunny** [comment]
```
git fetch origin
git rebase master
... resolve the conflict
git add <conflict file>
git rebase --continue
git push --force <your repo and branch>
```

**Comment by willemvd** [comment]
Will that work even in this broken state?

**Comment by willemvd** [comment]
@lunny created a new branch of upstream/master and cherry-picked the commits on top of it, after git push --force it seems to be ok now 😄

File: models/error.go [hunk 3]
```diff
@@ -887,3 +887,28 @@ func IsErrExternalLoginUserNotExist(err error) bool {
...
 	return fmt.Sprintf("external login user link does not exists [userID: %d, loginSourceID: %d]", err.UserID, err.LoginSourceID)
 }
+
+//  ________      _____   ____ _________________ ___
+//  \_____  \    /  _  \ |    |   \__    ___/   |   \
+//   /   |   \  /  /_\  \|    |   / |    | /    ~    \
+//  /    |    \/    |    \    |  /  |    | \    Y    /
+//  \_______  /\____|__  /______/   |____|  \___|_  /
+//          \/         \/                         \/
+
```
Comment (strk) [inline models/error.go]: Maybe you could take the chance to start with splitting error.go into error_oauth.go ...

Comment (willemvd) [reply models/error.go]: done , also moved some stuff from login_source.go to oauth2.go

**Comment by morrildl** [comment]
I just noticed this is on the 1.2.0 milestone; is there any chance it could make it into 1.1.0, or is it too late for that?

(I'm waiting on Google OAuth2 support before deploying a Gitea instance, and I'm impatient. :)

**Comment by lunny** [comment]
It should be in v1.2. You can use master version after this merged.

**Comment by willemvd** [comment]
Annoying... every time someone changes something to login_source.go it becomes a merge conflict since part of the code is moved to another file and it doesn't recognizes that quite right..

So I will wait for 1.1 to be released and 1.2 features gets merged to master

**Comment by strk** [comment]
@willemvd any chance to get dynamic client registration for `OpenID Connect` provider, in a way to let people log-in using an URL ? (seeking provider data in the page at URL and dynamically registering the client...)

**Comment by strk** [comment]
For dynamic client registration, see https://github.com/markbates/goth/issues/142

**Comment by willemvd** [comment]
@strk no don't have any plans to do so
Also feel a major drawbacks to this, because I think it will open up a possible security hole.
An admin is now not in control about what systems are attached and allows people from a possible not secure source.

Also think there is no real need since it is only supported by a very few providers which all allow manual registration (the current approach)

**Comment by strk** [comment]
What's an "insecure source" when it comes to authenticating users ?
If you have an "open registration" policy why "certainty" of identity
cannot be left to the user being authenticated ?

It's like letting users choose their passwords...

**Comment by morrildl** [comment]
FWIW, I agree with @willemvd.

The concern isn't so much that an incoming self-registered user identity is "false" (since the whole point of OpenID/OAuth is that you are delegating that), but rather that it can be stolen.

If you operate a Gitea instance with an open registration policy, you are presumably doing so to allow anyone to publish code for others to use. So, if I as a user choose to use a poor/insecure OpenID service to sign up, someone could steal my identity and use it to publish malware into repos I've published.

This is something that can happen with any provider, but it's far less likely with Github or Google or Facebook, than a random self-hosted OpenID instance or similar. It should be up to admins to decide which they trust. And while you can argue that the admins are choosing to trust dynamic registrations, it seems to me that Gitea as a tool for distributing source ought to be careful about what it enables.

IMO it makes sense to first finish up OAuth2 in general with a fixed list of supported providers, and get some deployments of that configuration. Then think about dynamic registrations later.

**Comment by strk** [comment]
I completely disagree on the assumption that a Big Auth Provider
gives more security. An attacker needs to only break into that Big Auth
Provider to compromise thousand of identities.

If someone stole your identity, I'll consider you responsible for
that and suspend your account, until you fix that (if I know and trust
you).

**Comment by willemvd** [comment]
Adding providers on the fly by Gitea user is basically adding users without passwords or security at all.
The only ones inGitea that should be able to add these kind of dynamic registration urls should be Gitea admins. That way they can skip the part of manually creating a secret and a key. That is the only reason why you should want this extra feature.

When Gitea end users can enter there own urls they could setup any open system on there provider side and always allow anyone.
OAuth2 is about delegating the login part to a trusted party and verify in the background that you agree both on the secret.
With dynamic registration from end users this is not possible in a secure way.

I'm not going to add this feature so if you (@strk) want this after all, you will have to submit a PR after this to add that.

**Comment by strk** [comment]
How's that different from letting users self-register ?
You still want some bot and spam protection, in both cases.

> When Gitea end users can enter there own urls they could setup any open system on there provider side and always allow anyone.

Same as creating a user and publishing its password, to let it be usable
by anyone.

I'm not saying to *force* that LoginSource to exist, just letting Admins
decide if they want or not to allow it.

Anyway: understood it's not in your plans to do that :)

**Comment by lunny** [comment]
@willemvd this could be continued since v1.1 released.

**Comment by lunny** [comment]
build failed.

File: models/login_source.go [hunk 4]
```diff
@@ -321,11 +329,25 @@ func GetLoginSourceByID(id int64) (*LoginSource, error) {
...
+		// keep track of the original values so we can restore in case of errors while registering OAuth2 providers
+		var err error
+		if originalLoginSource, err = GetLoginSourceByID(source.ID); err != nil {
+			return err
+		}
+	}
+
 	_, err := x.Id(source.ID).AllCols().Update(source)
-	if err == nil && source.IsOAuth2() {
+	if err == nil && source.IsOAuth2() && source.IsActived {
```
Comment (pgaskin) [inline models/login_source.go:342]: Spelling: source.IsActive , this needs to be renamed.

File: models/oauth2.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	"gitlab": {Name: "gitlab", DisplayName: "GitLab", Image: "/img/auth/gitlab.png",
+		CustomURLMapping: &oauth2.CustomURLMapping{
+			TokenURL:   oauth2.GetDefaultTokenURL("gitlab"),
+			AuthURL:    oauth2.GetDefaultAuthURL("gitlab"),
+			ProfileURL: oauth2.GetDefaultProfileURL("gitlab"),
+		},
+	},
+	"gplus":         {Name: "gplus", DisplayName: "Google+", Image: "/img/auth/google_plus.png"},
+	"openidConnect": {Name: "openidConnect", DisplayName: "OpenID Connect", Image: "/img/auth/openid_connect.png"},
+	"twitter":       {Name: "twitter", DisplayName: "Twitter", Image: "/img/auth/twitter.png"},
```
Comment (pgaskin) [inline models/oauth2.go:44]: Maybe add stack exchange as well?

File: templates/admin/auth/new.tmpl [hunk 6]
```diff
@@ -74,14 +74,37 @@
...
-		<div class="ui attached segment">
-			<h5>GMail Settings:</h5>
-			<p>Host: smtp.gmail.com, Port: 587, Enable TLS Encryption: true</p>
-			<h5>OAuth GitHub:</h5>
-			<p>{{.i18n.Tr "admin.auths.tip.github"}}</p>
+				<h4 class="ui top attached header">
+					{{.i18n.Tr "admin.auths.tips"}}
+				</h4>
+				<div class="ui attached segment">
+					<h5>GMail Settings:</h5>
```
Comment (pgaskin) [inline templates/admin/auth/new.tmpl]: Gmail, not GMail. Maybe also add this to translations?

File: templates/admin/auth/new.tmpl [hunk 7]
```diff
@@ -74,14 +74,37 @@
...
+						<span>{{.i18n.Tr "admin.auths.tip.gitlab"}}</span>
+						<li>Google+</li>
+						<span>{{.i18n.Tr "admin.auths.tip.google_plus"}}</span>
+						<li>OpenID Connect</li>
+						<span>{{.i18n.Tr "admin.auths.tip.openid_connect"}}</span>
+						<li>Twitter</li>
+						<span>{{.i18n.Tr "admin.auths.tip.twitter"}}</span>
+					</div>
+				</div>
+			</div>
```
Comment (pgaskin) [inline templates/admin/auth/new.tmpl]: Is there a way to make this automated (maybe a loop)?

File: models/login_source.go (hunk 4 above)
Comment (willemvd) [reply models/login_source.go:342]: This is reuse of existing code, so I think out of scope for this PR

File: models/oauth2.go (hunk 5 above)
Comment (willemvd) [reply models/oauth2.go:44]: Think this is not yet available in https://github.com/markbates/goth/#supported-providers or is it named differently?

File: templates/admin/auth/new.tmpl (hunk 6 above)
Comment (willemvd) [reply templates/admin/auth/new.tmpl]: This is reuse of existing code, so I think out of scope for this PR

File: templates/admin/auth/new.tmpl (hunk 7 above)
Comment (willemvd) [reply templates/admin/auth/new.tmpl]: Will have a look if this can be a loop over the existing OAuth2 providers

File: models/oauth2.go (hunk 5 above)
Comment (strk) [reply models/oauth2.go:44]: for the record; stackexchange (also?) supports OpenID-2.0, and you can use `openid.stackexchange.com` to login (will be converted to actual URL upon logging in)

File: models/login_source.go (hunk 4 above)
Comment (strk) [reply models/login_source.go:342]: Not only that, it's also in a DB model so changing that field would require a migration :/

**Comment by strk** [comment]
It would also be nice to get #1300 merged before this. @geek1011 can you resolve conflicts there ?

**Comment by lunny** [comment]
@willemvd is this ready to review?

**Comment by willemvd** [comment]
@lunny yes it is, unfortunately the last push didn't triggered a drone build so maybe you can manually trigger this (or I could do a merge/rebase with master)

**Comment by lunny** [comment]
@willemvd please send a rebase, I cannot find the link to the drone build.

**Comment by willemvd** [comment]
@lunny drone is almost done without failures, so it is ready! 😄

**Comment by lunny** [comment]
I have tested `google+` and `twitter`, both are OK. but `/user/link_account` UI is broken. Otherwise LGTM

**Comment by bkcsoft** [comment]
Are we allowed to use these logos? Why not use the ones in FontAwesome?

(Edit: I know for a fact that GitLab is missing from FA...)

File: routers/admin/auths.go [hunk 8]
```diff
@@ -7,16 +7,16 @@ package admin
...
-
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/auth"
 	"code.gitea.io/gitea/modules/auth/ldap"
+	"code.gitea.io/gitea/modules/auth/oauth2"
 	"code.gitea.io/gitea/modules/base"
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"github.com/Unknwon/com"
```
Comment (bkcsoft) [inline routers/admin/auths.go]: Newline before external packages

File: templates/admin/auth/edit.tmpl [hunk 9]
```diff
@@ -140,31 +140,64 @@
...
-						<label>{{.i18n.Tr "admin.auths.oauth2_provider"}}</label>
-						<div class="ui selection type dropdown">
-							<input type="hidden" id="oauth2_provider" name="oauth2_provider" value="{{$cfg.Provider}}" required>
-							<div class="text">{{.CurrentOAuth2Provider.DisplayName}}</div>
-							<i class="dropdown icon"></i>
-							<div class="menu">
-								{{range $key, $value := .OAuth2Providers}}
-									<div class="item" data-value="{{$key}}">{{$value.DisplayName}}</div>
-								{{end}}
+						<!-- OAuth2 -->
```
Comment (bkcsoft) [inline templates/admin/auth/edit.tmpl]: Weird indentation

**Review by bkcsoft — CHANGES_REQUESTED** [review]
formatting

File: routers/admin/auths.go (hunk 8 above)
Comment (willemvd) [reply routers/admin/auths.go]: added

File: templates/admin/auth/edit.tmpl (hunk 9 above)
Comment (willemvd) [reply templates/admin/auth/edit.tmpl]: fixed

**Comment by sapk** [comment]
Build seems to failed because the vendor `github.com/mrjones/oauth` doesn't correspond to remote. If you make change in the lib please revert them. A simple command to revert to original version is `govendor fetch github.com/mrjones/oauth`. This will also update the lib but, in this case, since it is not allready in repo there is no problems.

**Review by bkcsoft — APPROVED** [review]

**Comment by willemvd** [comment]
Didnt change anything to that lib so not sure why this happened, but I have updated the lib

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
LGTM

**Pull request merged**